### DISCUSSION TITLE
Flex Utils and other Flex pure AS3 Classes - Adding some of them to AIR?

### DISCUSSION THREAD

**Discussion by 2jfw** [airsdk/Adobe-Runtime-Support#3624] [Ideas]
I came across the need to use Base64 encoding and decoding and copied the existing Flex mx classes over to my project (/to a sub project for this, to be more specific).

1) https://github.com/apache/flex-sdk/blob/develop/frameworks/projects/framework/src/mx/utils/Base64Encoder.as
2) https://github.com/apache/flex-sdk/blob/develop/frameworks/projects/framework/src/mx/utils/Base64Decoder.as

Having developed Flex for over 10yrs now (still doing btw), I came across a couple of intersting _pure_ AS3 classes, which are only part of the Flex SDK and was wondering, if the AIR SDK may benefit from moving (/copying) a couple of them over.

I'm creating this thread to gather some feedback (pros/cons) on this and to create a general place for discussing this idea.
Also, so that others can point out more possibly interesting classes (utils/helpers) etc.

**Here a list of the Utils:**
https://github.com/apache/flex-sdk/tree/develop/frameworks/projects/framework/src/mx/utils

As pointed out, it's not all about the utils from Flex; there are other (pure as3) "candiates", which _may_ be a welcome addition to AIR, like some of the Validators for example (see https://github.com/apache/flex-sdk/tree/develop/frameworks/projects/framework/src/mx/validators).

Just wondering what others think of that?

**Comment by ajwfrost** [comment]
FYI base64 encoding/decoding can be done via the C++ implementations of [ByteArray](https://airsdk.dev/reference/actionscript/3.0/flash/utils/ByteArray.html) in AIR 51 - e.g. `createFromBase64` or `writeBase64`, and `toBase64` (similarly, base16 ("HexString") implementations were added too).

But I'm not quite sure whether you mean we should just add some of these (in source and/or swc format) into the AIR SDK, or whether we should update the built-in AS3 APIs to include some of these? (the latter may be a bit trickier to handle around licensing..)

thanks

**Reply by 2jfw** [reply]
Ah, many thanks for pointing out the native Base64 capabilites of ByteArray; I must have missed them. Will check this out and probably replace the mx classes then.

My thoughts about some of the mx utils/helpers/validators were, if it would benefit AIR to have some of those and to make them an actual part of the AIR SDK (e.g. by embedding them into the AIR runtime).
Similar to e.g. the existing AIR utils like `flash.globalization.CurrencyFormatter`, `flash.globalization.DateTimeFormatter` or `flash.globalization.NumberFormatter`.
There may be some more interesting mx classes which could add some useful functions to the AIR SDK.


**Side note:**
I came across the Base64Encoder from mx as I was typing that into the IDE to see if anything available popups up and to my suprise it actually was the `gamepad.swc`, which contains several of mx classes.
I would think these mx classes here are likely used to provide functionality, which the native AIR/flash classes where not able to provide at that point in time.

![grafik](https://github.com/user-attachments/assets/5dd90e18-53dd-4684-8c64-5134213c03b7)

**Comment by marchbold** [comment]
There is the as3corelib that was created as something like this idea by @mikechambers . It contains some core library utilities that have been pulled from various places. 

https://github.com/airsdk/as3corelib

Probably would be a good spot to add anything like this if needed

**Reply by 2jfw** [reply]
Ah, I somehow never looked into the as3commons package in detail and already forgot about it...
I will browse the classes to see if there are some additional interesting helpers or utils.
Thanks for pointing out!
